### PR TITLE
auth: add metrics to token cache

### PR DESCRIPTION
**Pull request by mikedanese**
Part of #81028

/sig auth
/kind feature

```release-note
NONE
```

@kubernetes/sig-auth-pr-reviews

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/authentication/token/cache/stats.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+const (
+	hitTag  = "hit"
+	missTag = "miss"
+
+	fetchActiveTag = "active"
+	fetchFailedTag = "error"
+	fetchOkTag     = "ok"
+)
+
+type statsCollector struct{}
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/authentication/token/cache/stats.go]: Why do we have this struct?

File: staging/src/k8s.io/apiserver/pkg/authentication/token/cache/stats.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+var (
+	requestLatency = metrics.NewHistogramVec(
+		&metrics.HistogramOpts{
+			Namespace:      "authentication",
+			Subsystem:      "token_cache",
+			Name:           "request_duration_seconds",
+			StabilityLevel: metrics.ALPHA,
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/authentication/token/cache/stats.go]: I assume this means we can change these metrics if we find that we made a mistake?

File: staging/src/k8s.io/apiserver/pkg/authentication/token/cache/stats.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+	start := time.Now()
+	return func(hit bool) {
+		var tag string
+		if hit {
+			tag = hitTag
+		} else {
+			tag = missTag
+		}
+
+		latency := time.Now().Sub(start)
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/authentication/token/cache/stats.go]: nit `latency := time.Since(start)`

**Review by enj — COMMENTED** [review]
Is there some process for "testing" these global metrics?

Generally LGTM

File: staging/src/k8s.io/apiserver/pkg/authentication/token/cache/stats.go (hunk 1 above)
Comment (mikedanese) [reply staging/src/k8s.io/apiserver/pkg/authentication/token/cache/stats.go]: Mostly for documentation in code. I can use a prefix like recordAuthenticating() on functions as well if you prefer that.

File: staging/src/k8s.io/apiserver/pkg/authentication/token/cache/stats.go (hunk 2 above)
Comment (mikedanese) [reply staging/src/k8s.io/apiserver/pkg/authentication/token/cache/stats.go]: Yes, I'm hoping to graduate them quickly once they are validated though.

File: staging/src/k8s.io/apiserver/pkg/authentication/token/cache/stats.go (hunk 3 above)
Comment (mikedanese) [reply staging/src/k8s.io/apiserver/pkg/authentication/token/cache/stats.go]: Done.

File: staging/src/k8s.io/apiserver/pkg/authentication/token/cache/stats.go (hunk 1 above)
Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/authentication/token/cache/stats.go]: Either is fine with me so we can leave as-is.

**Comment by enj** [comment]
/lgtm
/retest

> Is there some process for "testing" these global metrics?

@mikedanese thoughts on ^ ?

**Comment by mikedanese** [comment]
@enj I don't know of a way to do it that wouldn't be fairly contrived. Open to suggestions.

File: staging/src/k8s.io/apiserver/pkg/authentication/token/cache/stats.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+			Name:           "request_duration_seconds",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"status"},
+	)
+	requestCount = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Namespace:      "authentication",
+			Subsystem:      "token_cache",
+			Name:           "request_count",
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/authentication/token/cache/stats.go]: Conventionally, these should all be `_total` as per [prometheus naming guidelines](https://prometheus.io/docs/practices/naming/).

File: staging/src/k8s.io/apiserver/pkg/authentication/token/cache/stats.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"status"},
+	)
+	fetchCount = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
+			Namespace:      "authentication",
+			Subsystem:      "token_cache",
+			Name:           "fetch_count",
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/authentication/token/cache/stats.go]: same

File: staging/src/k8s.io/apiserver/pkg/authentication/token/cache/stats.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"status"},
+	)
+	blockCount = metrics.NewGauge(
+		&metrics.GaugeOpts{
+			Namespace:      "authentication",
+			Subsystem:      "token_cache",
+			Name:           "block_count",
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/authentication/token/cache/stats.go]: same

**Review by logicalhan — COMMENTED** [review]
/sig instrumentation

File: staging/src/k8s.io/apiserver/pkg/authentication/token/cache/stats.go (hunk 6 above)
Comment (mikedanese) [reply staging/src/k8s.io/apiserver/pkg/authentication/token/cache/stats.go]: This isn't a total though. It's not accumulating. It's the current number blocked. This is true for some of the guages in fetch_count as well. request_count is the only one that is accumulating.

Comment (logicalhan) [reply staging/src/k8s.io/apiserver/pkg/authentication/token/cache/stats.go]: Ah yeah that convention doesn't apply for gauges, only for the counters.

**Pull request merged**